oeps/architectural-decisions/oep-0065/decisions/0003-frontend-projects.rst

```diff
@@ -0,0 +1,143 @@
+* The approach of adding a config file and customizations to the source code is at odds with industry best practices and developer expectations when working with nearly any popular framework, such as Django, React, Next.js, or Rails.
+
+Finally, our documentation and best practices around how to build and customize the Open edX frontend have always been a bit of an afterthought, and operators have been left to "figure it out" and rely on tools like Tutor to manage the complexity and uncertainty for them. We believe that the platform itself should take more responsibility for providing best in class tools and patterns for working with the code.
+
+Decision
+********
+
+We will create a well-supported, built-in configuration and customization framework by adding support for frontend :term:`Projects <Project>`.
+
+A frontend Project is a repository, created and owned by an operator, responsible for containing all of the frontend configuration and customized modules for an operator's various environments.
+
+There are two primary types of projects: *site projects* and *module projects*.
+
+Site Projects
+=============
+
+A site project uses a :term:`Site Config` file to define the configuration for a :term:`Site`, which is an independently deployed portion of the Open edX frontend which renders a header, footer, and loads one or more Modules.
+
+A site project consists of:
+
+* One or more configuration files which account for all of an operator's config and customizations across all their environments.
+* A set of build targets expressed as ``"scripts"`` in ``package.json`` which point at ``openedx`` CLI commands from ``frontend-base``.
+* (Optional) A ``src`` sub-folder containing the operator's custom modules and extensions.
+
+A site project can load modules in several different ways (:term:`Importing <Imported Module>`, :term:`Federating <Federated Module>`, or :term:`Linking <Linked Module>`), as defined in its Site Config. The simplest is to ``import`` the module into the Site Config file so that it is bundled with the Site Project.
+
+.. image:: ../site-project-architecture.png
+
+To describe the steps in the above image:
+
+1. A build is started with the ``npm run build`` command, which references a ``scripts`` entry in the project's package.json.
+2. That script delegates to the ``openedx`` CLI ``build`` command provided by ``frontend-base``.
+3. The ``build`` CLI command runs webpack with the ``build`` webpack config.
+4. Webpack uses the :term:`Shell` - in ``frontend-base`` - as it's entry point.
+5. The Shell initialization code imports the :term:`Site Config` (i.e., ``site.config.build.tsx``) file from the project.
+6. The :term:`Site Config` file imports any :term:`Application Modules <Application Module>` from libraries it depends on (defined as ``dependencies`` in package.json), along with any other Modules from the ``src`` sub-folder.
+
+Module Projects
```

Comment:
While Site Projects are clear to me off the bat, I'm wondering why we also need module projects.   I suspect this is a consequence of the different ways modules can be built, but I was under the impression different package.json build targets in the modules themselves would be enough.

Reply:
I'm putting a clarifying paragraphs about why the two exist at the top of their sections, which hopefully clears it up a bit.